iree/compiler/Dialect/VM/Conversion/VMToEmitC/ConvertVMToEmitC.cpp

```diff
@@ -209,6 +237,389 @@ class GlobalStoreOpConversion : public OpConversionPattern<StoreOpTy> {
+        /*callee=*/rewriter.getStringAttr("VM_RETURN_IF_LIST_NULL"),
+        /*args=*/ArrayAttr{},
+        /*templateArgs=*/ArrayAttr{},
+        /*operands=*/ArrayRef<Value>{listDerefOp.getResult(0)});
+
+    auto getValueOp = failableCall(
+        /*rewriter=*/rewriter,
+        /*location=*/loc,
+        /*callee=*/rewriter.getStringAttr("iree_vm_list_get_value_as"),
+        /*args=*/
+        ArrayAttr::get(ctx, {rewriter.getIndexAttr(0), rewriter.getIndexAttr(1),
+                             StringAttr::get(ctx, valueTypeEnum.getValue()),
+                             rewriter.getIndexAttr(2)}),
+        /*templateArgs=*/ArrayAttr{},
+        /*operands=*/
+        ArrayRef<Value>{listDerefOp.getResult(0), getOp.index(),
+                        valuePtrOp.getResult()});
+
+    rewriter.replaceOpWithNewOp<emitc::CallOp>(
+        /*op=*/getOp,
+        /*type=*/getOp.getType(),
+        /*callee=*/rewriter.getStringAttr(valueExtractor.getValue()),
+        /*args=*/ArrayAttr{},
+        /*templateArgs=*/ArrayAttr{},
+        /*operands=*/ArrayRef<Value>{valuePtrOp.getResult()});
+
+    return success();
+  }
+};
+
+template <typename SetOpTy>
+class ListSetOpConversion : public OpConversionPattern<SetOpTy> {
+  using OpConversionPattern<SetOpTy>::OpConversionPattern;
+
+ private:
+  LogicalResult matchAndRewrite(
+      SetOpTy setOp, ArrayRef<Value> operands,
+      ConversionPatternRewriter &rewriter) const override {
+    auto ctx = setOp.getContext();
+    auto loc = setOp.getLoc();
+
+    Optional<StringRef> valueConstructor =
+        TypeSwitch<Operation *, Optional<StringRef>>(setOp.getOperation())
+            .Case<IREE::VM::ListSetI32Op>(
+                [&](auto op) { return StringRef("iree_vm_value_make_i32"); })
+            .template Case<IREE::VM::ListSetI64Op>(
+                [&](auto op) { return StringRef("iree_vm_value_make_i64"); })
+            .Default([](Operation *) { return None; });
+
+    if (!valueConstructor.hasValue()) {
+      return setOp.emitOpError() << " not handeled";
+    }
+
+    auto valueOp = rewriter.create<emitc::CallOp>(
+        /*location=*/loc,
+        /*type=*/emitc::OpaqueType::get(ctx, "iree_vm_value_t"),
+        /*callee=*/rewriter.getStringAttr(valueConstructor.getValue()),
+        /*args=*/ArrayAttr{},
+        /*templateArgs=*/ArrayAttr{},
+        /*operands=*/ArrayRef<Value>{setOp.value()});
+
+    auto valuePtrOp = rewriter.create<emitc::GetAddressOfOp>(
+        /*location=*/loc,
+        /*result=*/emitc::OpaqueType::get(ctx, "iree_vm_value_t*"),
+        /*operand=*/valueOp.getResult(0));
+
+    auto refOp = rewriter.create<emitc::CallOp>(
+        /*location=*/loc,
+        /*type=*/emitc::OpaqueType::get(ctx, "iree_vm_ref_t"),
+        /*callee=*/rewriter.getStringAttr("*"),
+        /*args=*/ArrayAttr{},
+        /*templateArgs=*/ArrayAttr{},
+        /*operands=*/ArrayRef<Value>{setOp.list()});
+
+    auto listDerefOp = rewriter.create<emitc::CallOp>(
```

Comment:
this is a good example of where emitc->emitc canonicalizations can be useful - all these derefs and such can CSE/fold away in a lot of cases before emitting code, which saves the C compiler processing the code from having to perform alias analysis and a bunch of other gnarly things that are unlikely to work right:

In the gist you posted for example v12/v13 and v15/v16 could fold away, as could the checks on them.
```c
v12 = *(v9);
v13 = iree_vm_list_deref(v12);
VM_RETURN_IF_LIST_NULL(v13);
v14 = iree_vm_list_reserve(v13, v2);
VM_RETURN_IF_ERROR(v14);
v15 = *(v9);
v16 = iree_vm_list_deref(v15);
VM_RETURN_IF_LIST_NULL(v16);
v17 = iree_vm_list_size(v16);
```
->
```c
v12 = *(v9);
v13 = iree_vm_list_deref(v12);
VM_RETURN_IF_LIST_NULL(v13);
v14 = iree_vm_list_reserve(v13, v2);
VM_RETURN_IF_ERROR(v14);
v17 = iree_vm_list_size(v13);
```

Which then can go even further for local values:
```c
v7 = &v6;
v8 = iree_vm_list_create(v5, v1, state->allocator, v7);
VM_RETURN_IF_ERROR(v8);
v9 = VM_GET_REF_ADDRESS(0);
v10 = iree_vm_list_type_id();
v11 = iree_vm_ref_wrap_assign(v6, v10, v9);
VM_RETURN_IF_ERROR(v11);
v12 = *(v9);
v13 = iree_vm_list_deref(v12);
VM_RETURN_IF_LIST_NULL(v13);
v14 = iree_vm_list_reserve(v13, v2);
```
->
```c
v7 = &v6;
v8 = iree_vm_list_create(v5, v1, state->allocator, v7);
VM_RETURN_IF_ERROR(v8);
v14 = iree_vm_list_reserve(v7, v2);
```

that's where the power of emitc will really shine - relying on the downstream compiler to do that same transformation with a bunch of those interleaved function calls to external code is never going to be as good as what you can do here :)

(<3 emitc :)